x/operator/keeper/abci.go

```diff
@@ -13,31 +13,53 @@ import (
 // operator according to the input assets filter and prices.
 func (k *Keeper) CalculateUSDValueForOperator(
 	ctx sdk.Context,
+	isForSlash bool,
 	operator string,
 	assetsFilter map[string]interface{},
 	decimals map[string]uint32,
 	prices map[string]operatortypes.Price,
-) (sdkmath.LegacyDec, sdkmath.LegacyDec, error) {
-	usdValue := sdkmath.LegacyNewDec(0)
-	selfUSDValue := sdkmath.LegacyNewDec(0)
+) (operatortypes.OperatorUSDValue, error) {
+	var err error
+	ret := operatortypes.OperatorUSDValue{
+		Staking:                 sdkmath.LegacyNewDec(0),
+		SelfStaking:             sdkmath.LegacyNewDec(0),
+		StakingAndWaitUnbonding: sdkmath.LegacyNewDec(0),
+	}
 	// iterate all assets owned by the operator to calculate its voting power
 	opFuncToIterateAssets := func(assetID string, state *assetstypes.OperatorAssetInfo) error {
-		price := prices[assetID]
-		decimal := decimals[assetID]
-		usdValue = usdValue.Add(CalculateUSDValue(state.TotalAmount, price.Value, decimal, price.Decimal))
-		// calculate the token amount from the share for the operator
-		selfAmount, err := delegationkeeper.TokensFromShares(state.OperatorShare, state.TotalShare, state.TotalAmount)
-		if err != nil {
-			return err
+		var price operatortypes.Price
+		var decimal uint32
+		if isForSlash {
+			// when calculated the USD value for slashing, the input prices map is null
+			// so the price needs to be retrieved here
+			price, err = k.oracleKeeper.GetSpecifiedAssetsPrice(ctx, assetID)
+			if err != nil {
+				return err
+			}
+			assetInfo, err := k.assetsKeeper.GetStakingAssetInfo(ctx, assetID)
+			if err != nil {
+				return err
+			}
+			decimal = assetInfo.AssetBasicInfo.Decimals
+			ret.StakingAndWaitUnbonding = ret.StakingAndWaitUnbonding.Add(CalculateUSDValue(state.TotalAmount.Add(state.WaitUnbondingAmount), price.Value, decimal, price.Decimal))
+		} else {
+			price = prices[assetID]
```

Comment:
Is it reliable to not check if prices map is nil?

Reply:
It may also be possible to remove the boolean and rely on whether the `map prices` is nil but I don't feel strongly about that.

Reply:
The flag is used to indicate different calculation ways for slash and the voting power update in `abci.go`, so using a flag is better.

Reply:
This is bad way in coding style, please always consider memory leak and never confuse the business logic with the coding logic, even theoretically it's not empty in business logic but still need to check in software design pattern

Reply:
> It may also be possible to remove the boolean and rely on whether the `map prices` is nil but I don't feel strongly about that.

Agree with Max

Reply:
I will add a detailed code comment for this function. It seems like it's the main reason for this issue.

Reply:
> This is bad way in coding style, please always consider memory leak and never confuse the business logic with the coding logic, even theoretically it's not empty in business logic but still need to check in software design pattern

It's not an issue related to coding style, and I don't see any risk of a memory leak. The core problem is the lack of detailed code comments to understand the business logic of this function.

Reply:
No, you messed the business logic with the coding design pattern, the business logic shouldn't destroy the coding regular rules!

Reply:
Although the check in the caller ensures the map isn't nil, it's better to check here. I will add a nil check here. As for removing the flag, I have added a detailed code comment to describe its usage, It's not a good idea to remove it.